Revert "{citra,yuzuPackages}: remove package"

Atemu opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

As explained in https://github.com/NixOS/nixpkgs/pull/293312#issuecomment-1994127648, an upstream source no longer being available does not warrant an immediate removal of a package that can build fine with cached sources for the time being.

Approval was only given by one of the 5 maintainers; the rest were not even notified. (This PR should ping them again.)

This is especially not an appropriate reaction when it is a package many people (including myself) used up until that point. Even if removal was the appropriate action, there should have been a grace period. 

Appropriate measures for this issue I can think of:

- Marking the package with a warning
- Using a mirror
- Setting yuzu.src.meta.broken = true 

Until such an appropriate measure is taken, revert to the previous status quo: A fully working Yuzu derivation with a minor problem for an extremely limited amount of users.

(If the maintainers do not wish to be listed here anymore because the package is "problematic" now, please speak out.)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Atemu committed:
Revert "citra: remove package"

This reverts commit 9b90f685e58c1483cbdea6388069e1ed8f57edae.

Atemu committed:
Revert "yuzuPackages: remove"

This reverts commit ef72e8b3a9b9743544b3ca6bca9da5770e20f51f.

K900 commented:
I would want to wait for the dust to settle a little bit, so we can be more confident this isn't going to lead to a DMCA strike or whatever.

Atemu commented:
That is my intention. Revert to the status quo, wait for an actual issue to crop up (i.e. DMCA notice) and then take a more extreme action.

I highly doubt Ninty would dare to attempt their corporate shitfuckery to this extent outside the US though.

thiagokokada commented:
I concur with @Atemu here. Let's wait, there is no reason to restore those packages until we have a proper upstream (that is also not DMCA'd).

For everyone that wants to use those packages, you can always just copy the derivation to your configuration and get a copy of the repo somewhere.

Atemu commented:
> there is no reason to restore those packages until we have a proper upstream

That's the point I disagree with. These packages have users and they work for them. We are actively removing functionality from their systems.

This was not a cleanup of dead or broken code; it still works. I just built yuzu right on my system. The source came from cache.nixos.org which is obviously not ideal but not the end of the world either.

I intended to keep this rather simple to unbreak users but switching out the sources for an alive mirror should be rather simple and would leave the yuzu package without any significant technical issue.

thiagokokada commented:
> That's the point I disagree with. These packages have users and they work for them. We are actively removing functionality from their systems.

Removal of packages or features will always break someone relying on that functionality. Heck, even a refactor made wrong will break some users.

In the end we need to decide between what is worth maintaining and what it is not, and in my opinion keeping those packages is not worth maintaining them until we get a new upstream that will not be DMCA'd next week.

> That's the point I disagree with. These packages have users and they work for them. We are actively removing functionality from their systems.

I think we agree to disagree at this point.

Scrumplex commented:
It should also be emphasized that the early-access source code provided by pineappleEA is still available and has been available despite upstream yuzu being taken down.

See https://github.com/pineappleEA/pineapple-src

thiagokokada commented:
> It should also be emphasized that the early-access source code provided by pineappleEA is still available and has been available despite upstream yuzu being taken down.
> 
> See https://github.com/pineappleEA/pineapple-src

It doesn't help that `pineappleEA` only publishes the code and doesn't seem interested in maintaining the code (that IMO is more important to me). Looking at the commit history the only commit after the whole mess was an update to the `README`.

I would be ok if `pineappleEA` is re-introduced as `pineapple-ea` instead of `yuzu-early-access` though. At least this would make it clear that this is not Yuzu, so there is no maintained upstream.

However I would still prefer to wait because I am sure eventually a actual upstream will appear.

thiagokokada reviewed (changes requested):
I am putting this as "request changes" since I don't concur with this decision, but if there is a consensus that we need to restore those feel free to dismiss my review.

Atemu commented:
> In the end we need to decide between what is worth maintaining and what it is not, and in my opinion keeping those packages is not worth maintaining them until we get a new upstream that will not be DMCA'd next week.

I fail to see any sort of maintenance burden here. I can assure you that the package will not have to account for upstream changes in the foreseeable future ;)

I'm willing to step up as a maintainer in the interim if you're at all worried.

I'm also again of the opinion that packages should not be removed on the grounds that there could potentially possibly be a burden sometime far into the future but rather at the point where an actual burden materialises.

> only publishes the code and doesn't seem interested in maintaining the code (that IMO is more important to me).

That's enough for them to function as a mirror which is all we need to mend the "pressing" technical issue of source unavailability.

Yes, upstream development is no longer taking place. If I had to guess the same is true for 1000s if not 10000s of packages in Nixpkgs; they're not hurting anyone.

thiagokokada commented:
> I fail to see any sort of maintenance burden here. I can assure you that the package will not have to account for upstream changes in the foreseeable future ;)

Maintenance burden also happens when we do any changes in dependencies that can break because a package depends in a older version of this package.

Generally who handles those fixes is upstream, but without an upstream who is going to do those? In lack of upstream doing the change (because e.g.: they don't support third-party packaging for example) we can do it ourselves, but generally having the source code available is crucial for those cases and right now we don't (unless you download the code from Hydra, that really is a PITA).

> That's enough for them to function as a mirror which is all we need to mend the "pressing" technical issue of source unavailability.

See point above.

AndersonTorres commented:
> That's the point I disagree with. These packages have users and they work for them. We are actively removing functionality from their systems.

Yes, because upstream died.

> This was not a cleanup of dead or broken code; it still works.

This is literally a dead and broken code, since the upstream repo disappeared.

> I just built yuzu right on my system. The source came from cache.nixos.org which is obviously not ideal but not the end of the world either.

1. This is not ideal, caching is a workaround. This is an argument for its removal, not for bit-rotting the repo.
2. There are projects that do the archaeological work you are proposing:

- https://lazamar.co.uk/nix-versions/
- https://www.nixhub.io/

> I intended to keep this rather simple to unbreak users but switching out the sources for an alive mirror should be rather simple and would leave the yuzu package without any significant technical issue.

Let the [NURers](https://github.com/nix-community/NUR) do it, then.

> I fail to see any sort of maintenance burden here. I can assure you that the package will not have to account for upstream changes in the foreseeable future ;)

Including security bugs (that will be discovered some day) that we will not patch (since we are mere packagers)?

This is precisely another argument for removal, not for keeping the corpse.

yu-re-ka reviewed (approved):
I am in favor of keeping around.

There is nothing that would make this criminal anyways, just like youtube-dl isn't criminal.

Indeed [SuYu](https://gitlab.com/suyu-emu/suyu) looks quite active already, and even if that project dies there is going to be ten new ones. I don't worry about maintenance burden falling on nixpkgs devs at all. Quite the opposite, this is one of the easier packages compared to the other corpses we have lying around (OpenCV2 anyone?).

AndersonTorres commented:
Are you arguing for packaging Suyu or for reverting the removal of Yuzu?

yu-re-ka commented:
I would suggest reverting the removal of Yuzu now, and switching the source over to SuYu (or any other fork) at a later point (when a suitable tagged version is available).

ElvishJerricco commented:
Inaccessible tarballs that only work by the grace of `cache.nixos.org` is not an acceptable thing to knowingly merge into nixpkgs IMO. Sure, there are probably a lot of packages that rely on this in nixpkgs but that's not on purpose and they should probably be removed when noticed. This is a knowing action to add broken packaging to nixpkgs.

If we were changing the package to point to a new fork or something, it'd be a different story.

JoshuaFern reviewed (changes requested):
This should not be reverted for reasons already mentioned.

I think with a little patience you'll see forks like Suyu in nixpkgs soon. In the mean time users still have plenty of ways to run Yuzu.

Atemu commented:
> Maintenance burden also happens when we do any changes in dependencies that can break because a package depends in an older version of this package.

Sure. If yuzu should actually break because of a deeply incompatible dep upgrade in a year or so, you can remove it.

(Not without asking its maintainers to make it compatible with reasonable notice beforehand of course.)

> who handles those fixes is upstream, but without an upstream who is going to do those?

...us? We're a distro. We make software compatible with the rest of the package set. That's our job. We patch software *all the time*.

There is absolutely no reason why we wouldn't be able to have compat patches. Many of us have created such patches without any upstream help before. It's quite common to have to patch around Nixpkgs-specific issues.  
You don't need a responsive upstream to do that (or at least I never did). Having one is preferable of course but not required. 

That's all of course *if that case were to even materialise*. We're talking about a scenario where:

1. A deep incompatibility is brought about in an update of a dep that affects yuzu
2. Yuzu is the only significant breakage
3. Yuzu maintainers are not able to patch around the issue
4. Maintainers of the dep somehow decide it's not worth breaking a known abandoned package

Then and only then would yuzu become a minor burden.

***

I do not feel like discussing this "argument" about the source being MIA being a critical issue any further. The only "argument" against it I've seen so far is blatant whatboutism. Are you going to ask me next what would happen if the bytes of the cached src on cache.nixos.org were to flip?

Yeah of course it's not great to not have the source available upstream. It's an actual issue that should be resolved.

It's not at all a critical one however. If the package breaks because of it in a month, so what? Let it break when that actually happens and don't artificially break it because it could *potentially* be broken in the future (maybe).

I don't cut my arm off to go skiing because I could potentially break it when I fall. That makes no sense at all.

Atemu commented:
> This is literally a dead and broken code, since the upstream repo disappeared.

The only dead code is the fetcher. I can replace it with a `requireFile` if you'd like.

> Including security bugs (that will be discovered some day) that we will not patch (since we are mere packagers)?

Why would we not patch security bugs when they're discovered?

Also, should that happen, the appropriate action would still not be immediate removal but simply marking the package as insecure.

If a known secuirty issue isn't fixed for months or years, only then I'd start thinking about removal.

Atemu commented:
> Inaccessible tarballs that only work by the grace of `cache.nixos.org` is not an acceptable thing to knowingly merge into nixpkgs IMO.

I can see where you're coming from but note that this is a revert of a PR that IMHO was merged prematurely.

Would you be comfortable with it if I simply replaced the dead fetchers with `requireFile`? I'll have to test whether substitutes work on those.

Alternatively, I might just take a mirror and call it a day as then there'd be no technical issue anymore.

AndersonTorres commented:
> I think with a little patience you'll see forks like Suyu in nixpkgs soon. In the mean time users still have plenty of ways to run Yuzu.

In the worst case, it would be an argument for keeping it at Unstable until the next cycle or next merging with Staging.
"Oh I have fond memories" is not a reason to keep the code until the end of times - as it is clearly the intention of OP (_oh only then I'd maybe start to thinking about the possibility of a well-studied and well-argued and well-embased grace period for removal..._).

> Sure. If yuzu should actually break because of a deeply incompatible dep upgrade in a year or so, you can remove it.
> (Not without asking its maintainers to make it compatible with reasonable notice beforehand of course.)

Like, huh, the dreadful Python 2?
Like, huh, that thread from 2022 about removing unmaintained software? https://github.com/NixOS/nixpkgs/pull/157959



> Why would we not patch security bugs when they're discovered?

Tell this to those who marked Python2 as insecure.

> Also, should that happen, the appropriate action would still not be immediate removal but simply marking the package as insecure.

Maybe it's time to implement https://github.com/NixOS/rfcs/pull/127...  

> ...us? We're a distro. We make software compatible with the rest of the package set. That's our job. We patch software all the time.

1. Up to which point updating patches does not become acting like an upstream maintainer?
   1. worse, keeping patches for a dead upstream based on a cached copy without any clue from original source history...
      1. yes I know about full backups of this code
1. We are not the only package manager in this planet. Keeping an inside fork fork of a software otherwise not related to the main activity of Nixpkgs defats its purpose.
    What if AUR wants to package Yuzu? They should rely on our caches and patches?

    The software we keep is called Nixpkgs, not Nixpkgs-and-what-is-on-Hydra-cache.

> It's quite common to have to patch around Nixpkgs-specific issues.

1. There is a huge difference between patching a software to make it run on NixOS because of our particular concerns (like, say, reproducibility), and patching a software to keeping it up-to-date.
1. We regularly offer general-purpose patches to upstream projects. We did it less than a week ago with [Jasper](https://github.com/NixOS/nixpkgs/pull/294994). 
1. Keeping such general-purpose patches around the code is not what we usually do in Nixpkgs. Indeed we `fetchpatch` - we literally fetch patches! - from various external sources.

> You cannot use graphical applications from revisions other than your system revision without terrible hacks.

Nonetheless, you suggest us the Nixpkgs ecosystem keep it indefinitely.
You give an argument for preservation with one hand, then you give an argument against preservation with the other hand.

> Then and only then would yuzu become a minor burden.

Minor burden? Are you seriously saying that keeping a C++ project that tries to emulate a computer system via a truckload of reverse engineering is "a minor burden"?

> If a known secuirty issue isn't fixed for months or years, only then I'd start thinking about removal.

What else? Maybe Pentagon being invaded because of this security bug will make you start to thinking about maybe re-reverting this PR?

> I have yet to see a good argument for how this "cleaning up" is going to help anyone.

"Not caring about code that the upstream is gone" helps a lot.

> Actually dead code that does not build anymore, where the package has no responsive maintainers and/or causes a concrete issue, yeah remove that shit.

Please decide your course of action!

With one hand, you argue about not keeping the package at its first miscompilation, while with the other hand you suggest to keep patching it indefinitely?

> Code that still works (I just built it), has maintainers

It does not have (upstream) maintainers. Have you read the news?

> Alternatively, I might just take a mirror and call it a day as then there'd be no technical issue anymore.

Your purpose, from what you wrote, is not merely reverting a premature PR, but keeping yuzu (citra? who cares?) forever.

AndersonTorres commented:
> Alternatively, I might just take a mirror and call it a day as then there'd be no technical issue anymore.

My math teacher usually said "I want to see such ideas implemented, not merely spoken".

Show us at least a GitHub/GitLab/Codeberg/whatever mirror you like and package it.

(Maybe we can rename it "uzuy"...)

thiagokokada commented:
> ...us? We're a distro. We make software compatible with the rest of the package set. That's our job. We patch software all the time.

Please don't quote a small part what I said out of context. I said that of course we can patch, but this has deeper implications them what you said.

> That's all of course if that case were to even materialise. We're talking about a scenario where:
>    1. A deep incompatibility is brought about in an update of a dep that affects yuzu
>    2. Yuzu is the only significant breakage
>    3. Yuzu maintainers are not able to patch around the issue
>    4. Maintainers of the dep somehow decide it's not worth breaking a known abandoned package

Point 1 happens all the time considering we are rolling release distro. Point 2 is irrelevant (Yuzu being the only breakage or part of a bigger breakage is not the issue here, is lacking upstream to help us fixing those issues the actual issue here). Point 3 forgets the how: without the source code we will need to download the source from Hydra, without the Git history, to fix the issue in many cases. Point 4 goes to the point 1, the rolling release part.

> I do not feel like discussing this "argument" about the source being MIA being a critical issue any further. The only "argument" against it I've seen so far is blatant whatboutism.

It is not the only argument, and also is not whatboutism. There are lots of arguments in both this thread and in the removal PR. You may not concur with my or other people arguments, this is fine, but no need to be sarcastic.

ElvishJerricco commented:
> I do not feel like discussing this "argument" about the source being MIA being a critical issue any further.

I'm sorry you don't feel like discussing it, because IMO you haven't addressed the problem. The code doesn't build. The good graces of `cache.nixos.org` happening to have the sources does not constitute a Nix expression that builds. I don't care about patches, or maintenance, or any of that other stuff. I care that it does not build. We should not be merging fundamentally broken expressions into nixpkgs.

Atemu commented:
> In the worst case, it would be an argument for keeping it at Unstable until the next cycle or next merging with Staging.

Staging cycles bear no significance for keeping or not keeping packages. Their one and only purpose is to group large rebuilds.

> Oh I have fond memories" is not a reason to keep the code until the end of times - as it is clearly the intention of OP

Please take your false dichotomies elsewhere and don't lay words in my mouth.

<details>

<summary>Only concerns the person I'm responding to</summary>

I've had less than great interactions with you in particular before and you're really close to getting a report to the mod team handed in about you because this is not how I wish to be treated in this community.  
I'd rather not have to do that and find a way to be nice to each other instead. Please watch it. **I do not appreciate trolling.**

</details>

> Like, huh, the dreadful Python 2?
> Like, huh, that thread from 2022 about removing unmaintained software? #157959

I do not see what python2 has to do with this?

Yuzu does not depend on python2 in any way that I'm aware of.

> Tell this to those who marked Python2 as insecure.

Sorry, I missed a word in the sentence: "Why would we not *attempt* to patch security bugs when they're discovered?"

It was attempted for python2. Python2 was marked insecure because it remained insecure despite attempts, with reasonable notice and attempted alternatives.

Its case is not at all comparable to this one however as it's a *very* core library rather than a leaf package. Though I'd still mark yuzu as insecure if a security issue was to be found at some point in the future that we would not be able to patch around.

> Up to which point updating patches does not become acting like an upstream maintainer?

I'd draw the line at developing new features related to the core of its purpose. Patching things to be compatible or secure is historically well within the domain of distros.

I don't see any merit discussing the philosophy of when someone is considered an upstream maintainer in this context.

> 2\. We are not the only package manager in this planet. Keeping an inside fork fork of a software otherwise not related to the main activity of Nixpkgs defats its purpose.
>         What if AUR wants to package Yuzu? They should rely on our caches and patches?

I know the merits of upstream sources full and well. I too would prefer a stable and active Yuzu upstream. We do not need to discuss this.

> 1. There is a huge difference between patching a software to make it run on NixOS because of our particular concerns (like, say, reproducibility), and patching a software to keeping it up-to-date.

Okay, point still stands. A quick ripgrep reveals 313 instances of `pythonRelaxDepsHook` to patch software to "keep up to date" with our newer deps. I believe many more instances of that exist done manually.

That's about the extent of "keeping up to date" we'd need to do for yuzu. (Note that this is an exclusive "we", you or anyone else uninterested in yuzu is not obliged in any way to touch yuzu in any way.)

> We regularly offer general-purpose patches to upstream projects. We did it less than a week ago with [Jasper](https://github.com/NixOS/nixpkgs/pull/294994).

I advocate for that too. Don't see how that applies here in any way as upstream is quite obviously dead.

> Keeping such general-purpose patches around the code is not what we usually do in Nixpkgs.

If by "general-purpose patches" you mean compatibility and security patches then that is a false statement. We have plenty of those in-tree.  
If not, I don't see why you'd mention this because patches other than compatibility and security were not up to discussion before.

> Nonetheless, you suggest us the Nixpkgs ecosystem keep it indefinitely.

I do not.

> Minor burden? Are you seriously saying that keeping a C++ project that tries to emulate a computer system via a truckload of reverse engineering is "a minor burden"?

I'm not too familiar with its packaging but I've seen far worse. You're missing the point entirely here though because this is about it not being a burden for maintainers of dependencies, not the package's maintainers.

> Maybe Pentagon being invaded because of this security bug will make you start to thinking about maybe re-reverting this PR?

Please make serious arguments. I do not appreciate trolling.

> "Not caring about code that the upstream is gone" helps a lot.

If you're not interested in yuzu, feel free to ignore it. Nobody is asking you to maintain yuzu in any way whatsoever.

If you happened to be a maintainer of a dep of yuzu (correct me if I'm wrong here but I don't believe you are), the absolute most you could be expected to do for a reverse dependency such as yuzu is to notify their package maintainers should their package break because of an update you did and allow them a little time to find a solution. Just like any other "significant" (as in: people care about and has active maintainers) reverse dependency.

> With one hand, you argue about not keeping the package at its first miscompilation, while with the other hand you suggest to keep patching it indefinitely?

I do not argue for either of those.

> It does not have (upstream) maintainers. Have you read the news?

Please note the difference between upstream maintainers and package maintainers.

I have not explicitly stated it because I thought it was obvious but at no point did I mean "upstream maintainers" when I said "maintainers". Upstream maintainers practically do not exist for the time being.

> Your purpose, from what you wrote, is not merely reverting a premature PR, but keeping yuzu (citra? who cares?) forever.

Ah yes of course, you know that so much better than me. Lol.

Atemu commented:
> Please don't quote a small part what I said out of context. I said that of course we can patch, but this has deeper implications them what you said.

Ah, I had assumed the first part of the sentence was the main point. I'll respond to the latter:

> > Generally who handles those fixes is upstream, but without an upstream who is going to do those? In lack of upstream doing the change (because e.g.: they don't support third-party packaging for example) we can do it ourselves, but generally having the source code available is crucial for those cases and right now we don't (unless you download the code from Hydra, that really is a PITA).

Mirrors exist for it everywhere. Good access to code is obviously desirable (certainly in the long term) but it's not like it's been banished off the face of the earth. If you really need it, it's easy to get. I'd also assume the active maintainers have a local copy anyways. It won't prevent anyone from creating patches.

> Maintenance burden also happens when we do any changes in dependencies that can break because a package depends in an older version of this package.

I have sympathy for this but in this specific case, if yuzu were to actually break because of a dependency, then that would be the time to remove it. (Not without giving its maintainers a chance to resolve the issues themselves of course.)  

It has known problems, nobody could expect outside maintainers to care much about the wellbeing of a package with known issues. We currently don't have the infrastructure for a standardised method of declaring it as such yet but we can still do so in other ways.

What would you think of a comment at the top of the file stating that the package does not have an alive upstream but is kept because it has active users and is subject to removal once it falls into disrepair? Effectively a sort of deprecation notice.

> Point 1 happens all the time considering we are rolling release distro.

It's a subtle distinction but being rolling release does not mean deps become incompatible ways "all the time" but that deps *could* become incompatible at any time rather than at specific times.

It still has to happen in the first place. It could happen at any time but that doesn't mean it's likely to happen short-term. Long-term, I'd agree; I could certainly see Yuzu breaking due to a dep upgrade in a year or two.  
But as mentioned previously, that does not warrant immediate removal; especially not in the current situation where there may be a suitable alternative upstream within a few weeks or months. Let's cross that bridge when we get there.

> Point 2 is irrelevant (Yuzu being the only breakage or part of a bigger breakage is not the issue here, is lacking upstream to help us fixing those issues the actual issue here). Point

Of course no upstream patches would be the underlying cause but the maintenance burden for dep maintainers would only materialise if yuzu was a significant breakage and they were somehow obliged to care about it.  
If you proposed a change to, say, glibc that broke half of our packages including yuzu, I would not for example say that yuzu would be a significant maintenance burden.  
If it only broke yuzu and a couple of effectively dead packages, then I'd agree though I'd still say that dep maintainers shouldn't feel obliged to care about yuzu in that case.

> Point 3 forgets the how: without the source code we will need to download the source from Hydra, without the Git history, to fix the issue in many cases

As mentioned, this is not the case. The git history didn't magically disappear off the internet.

> It is not the only argument, and also is not whatboutism. There are lots of arguments in both this thread and in the removal PR.

I was specifically talking about arguments against the source being MIA somehow being a *critical issue*. Sorry for not making that clear enough.  
I've yet to see an argument that did not boil down to whataboutism. If I missed one, let me know.

Atemu commented:
I had intended for this to become a simple reversion to the status quo and was not expecting to be expected to bring improvements over the status quo but I don't mind doing these.

So for progress on this PR:

* I'll substitute the source URLs with suitable mirrors. Hashes will stay the same.
* I'll add myself as maintainer for the time being because clearly people think that the previous maintainers would somehow not be able to deal with this package's "critical" issue. Ping me in case Yuzu causes you any significant burden at all in your Nixpkgs contributions.

Atemu committed:
yuzu, citra: use mirrors for src

This solves the technical issue of upstream sources being unavailable. Which
doesn't block the build because the source is cached on cache.nixos.org but
breaks an aspect of reproducibility because others can no longer reproduce the
source without the binary cache.

As you can see, the hashes are all the same; no content has changed.

The implementation is quite complex because fetchgit isn't made to use mirrors
for submodules and submodules are also just kind of a pain to deal with in
general.

(The code could be deduplicated but it's expected to be obsolete in a few
month's time.)

yuzu-ea did not have to be touched; its source is still available.

Atemu committed:
yuzu, citra: add notice

Atemu committed:
yuzu, citra: add myself as maintainer for the interim

Atemu commented:
I did the thing and all sources are now reproducible again. It's not pretty but it works and is good enough.

I also added the notice and made myself the maintainer of these packages for the interim. I'll again extend my offer to the previous maintainers to remove them if they no longer wish to maintain these packages.

Atemu committed:
yuzu, citra: remove update scripts

Won't be getting updates any time soon :(

JoshuaFern commented:
I'm unsubscribing from this discussion as well. I believe this entire proposal is driven by some kind of emotional desire not to give up on Yuzu, which I empathize with.

My stance is that this change should not be reverted, `cache.nixos.org` or mirrors should not be used, and whatever other "solutions" and arguments are manifested to include broken, dead, and unmaintained software into nixpkgs should be rejected.

AndersonTorres commented:
> Staging cycles bear no significance for keeping or not keeping packages. Their one and only purpose is to group large rebuilds.

Since Staging brings too many mass-rebuild packages to Master, it has bigger probability of clash.
But this is not relevant to the arguments in any way.

Any "mirror" of the most recent Yuzu code falls under the first condition for "not being accepted" under the packaging guidelines

> Is the package ready for general use? We don't want to include projects that are too immature or are going to be **abandoned** immediately. In case of doubt, **check with upstream**. [link](https://github.com/NixOS/nixpkgs/tree/master/pkgs#quick-start-to-adding-a-package) 

Bold mine. 
Further, remembering that I helped writing this in order to be more welcoming to new contributors. The original proposal was way harsher.

> Please take your false dichotomies elsewhere and don't lay words in my mouth.

This is precisely the tone you are directing this pull request: keeping a dead-upstream package at all costs.

> I do not see what python2 has to do with this? Yuzu does not depend on python2 in any way that I'm aware of.

I gave two examples of similar polemics of keeping a software after its EOL.

> Sorry, I missed a word in the sentence: "Why would we not _attempt_ to patch security bugs when they're discovered?"

We can anything we like (given that it is not a sin or a crime or a trigger for a Nintendo lawsuit).
But this is not our _duty_ as package maintainers.

> 
> It was attempted for python2. Python2 was marked insecure because it remained insecure despite attempts, with reasonable notice and attempted alternatives.

Even after _reasonable notice and attempted alternatives_, Profpatsch loudly complained about their packages being removed.

Looks like the story is repeating.

> 
> > Up to which point updating patches does not become acting like an upstream maintainer?
> 
> I'd draw the line at developing new features related to the core of its purpose. Patching things to be compatible or secure is historically well within the domain of distros.

We have at least two problems here.

- Yuzu is not meant to be a static piece of code that will never change. Indeed such emulators need to deal with new softwares and new functionalities created for the original hardware (and the host hardware too). In other words, Yuzu is obsoleted software since 2024-03-04.
- Even before that "only security or otherwise minor patches" line, the ideal scenario should be uploading those general-purpose patches to the upstream. And, as I said before, upstream is gone and backup repos are at most a subpar band-aid.

This expression is being kept with the help of too much devices.

> I don't see any merit discussing the philosophy of when someone is considered an upstream maintainer in this context.

As I said before, maintaining a collection of patches for a backup of a dead upstream is not our duty.

> Okay, point still stands. A quick ripgrep reveals 313 instances of `pythonRelaxDepsHook` to patch software to "keep up to date" with our newer deps. I believe many more instances of that exist done manually.

Not going into the intricacies of Python projects,
As I said before, in an ideal scenario, those patches and hacks should be upstreamed.

> That's about the extent of "keeping up to date" we'd need to do for yuzu. (Note that this is an exclusive "we", you or anyone else uninterested in yuzu is not obliged in any way to touch yuzu in any way.)

This "exclusive we" is an argument for keeping this in a NUR, not in the Nixpkgs monorepo.

> I advocate for that too. Don't see how that applies here in any way as upstream is quite obviously dead.

And this is the master point of the original pull request you are now reverting.

> If by "general-purpose patches" you mean compatibility and security patches then that is a false statement. We have plenty of those in-tree. If not, I don't see why you'd mention this because patches other than compatibility and security were not up to discussion before.

Point taken. 
Nonetheless this is not the ideal - again and again, it is not our duty to be a repository of general-purpose patches, they should be offered to upstream.

> I do not.

This is the tone of your whole banzé. Your arguments always impose a series of "maybe" steps before removal of this dead upstream.

> Please make serious arguments. I do not appreciate trolling.

You are being sarcastic and passive-aggressive all the time. "Pot meets kettle", as they say in Anglophone world? LOL

<details>
<summary><i>Answering the hidden content</i></summary>

It's funny when you rely on passive-agressive behavior, whereas when I retort in the same tone pointing out the  ridiculousness of your "keep it indefinitely" proposal, your reaction is a threat to call the mods!

Maybe I should need to learn how to be more sensitive to sarcasm and threaten to call the mods when I am the target of one!

</details>

> If you're not interested in yuzu, feel free to ignore it. Nobody is asking you to maintain yuzu in any way whatsoever.

Again, another argument for keeping it in a NUR.

> If you happened to be a maintainer of a dep of yuzu (correct me if I'm wrong here but I don't believe you are)

Just to be sure, Yuzu depends on Qt and [CMake](https://github.com/NixOS/nixpkgs/blob/057f9aecfb71c4437d2b27d3323df7f93c010b7e/pkgs/by-name/cm/cmake/package.nix#L184). 

It will be impossible to _ignore it_ when/if they become incompatible - yet another reason to NURing it.

Further, personalizing arguments this way is a little bit dishonest. 
I don't like it. I hate badge-flashing arguments.

Not that you care, seemingly...

> Ah yes of course, you know that so much better than me. Lol.

Paraphrasing myself:

Your purpose, from what you wrote - **from what you wrote** -, is not merely reverting a premature PR, but keeping this packages indefinitely.

I can't read your mind, I can only read your words.

Aleksanaa commented:
yuzu release is out, let's move to it

K900 commented:
I would give it some more time. The 0.0.2 release of suyu is ... extremely messy.

aanderse commented:
hi team :wave:

where are we on this? new release coming up, it would be nice to have something easily available...

K900 commented:
Yuzu is very dead.

aanderse commented:
i think everyone here understands that - so what will we do? package `suyu`? abandon everything all together and add an article to the wiki with a `nix` expression on how to build the last release of `yuzu` locally? or... ?

K900 commented:
Recommend people to use Ryujinx. Suyu is effectively also dead.

AndersonTorres commented:
Last month:

https://git.suyu.dev/suyu/suyu/releases/tag/v0.0.3

nyabinary commented:
> Recommend people to use Ryujinx. Suyu is effectively also dead.

What about for 3ds emulation?

K900 commented:
We can see which of the forks is alive, if any, and package that.

AndersonTorres commented:
Lime3ds, I believe...

nyabinary commented:
> Lime3ds, I believe...

Panda3DS too (not a fork but an independent 3ds emulator)

nyanloutre commented:
Hello,

I tried to create a generic framework to build yuzu forks in this PR to make it easier to support new forks : #403752
For a starting point I tried to build forks listed as active in [this wiki](https://emulation.gametechwiki.com/index.php/Yuzu#Forks) and the new eden fork that seems to gain traction and has good performances.

Would anyone active in this PR give me some feedbacks ?
Thanks,

K900 commented:
I don't think we should be putting this into nixpkgs, especially until there's consistent momentum behind one of the forks. Maybe put it into a separate flake for now?

pbsds commented:
* https://github.com/NixOS/nixpkgs/pull/406445

thiagokokada commented:
Closing this PR since we have alternatives now (even if the consensus seems that we also need to drop the alternatives).
